src/main/java/appeng/util/helpers/ItemComparisonHelper.java

```diff
@@ -0,0 +1,443 @@
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Applied Energistics 2.  If not, see <http://www.gnu.org/licenses/lgpl>.
+ */
+
+package appeng.util.helpers;
+
+
+import java.lang.reflect.Field;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+import javax.annotation.Nullable;
+
+import net.minecraft.item.Item;
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.NBTBase;
+import net.minecraft.nbt.NBTPrimitive;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.nbt.NBTTagList;
+import net.minecraft.nbt.NBTTagString;
+import net.minecraftforge.oredict.OreDictionary;
+
+import appeng.api.config.FuzzyMode;
+import appeng.core.AELog;
+import appeng.util.item.AESharedNBT;
+import appeng.util.item.OreHelper;
+import appeng.util.item.OreReference;
+
+
+/**
+ * A helper class for comparing {@link Item}, {@link ItemStack} or NBT
+ *
+ */
+public class ItemComparisonHelper
+{
+
+	private static Field tagList;
+
+	/**
+	 * Compare the two {@link ItemStack}s based on the same {@link Item} and damage value.
+	 * 
+	 * In case of the item being damageable, only the {@link Item} will be considered.
+	 * If not it will also compare both damage values.
+	 * 
+	 * Ignores NBT.
+	 * 
+	 * @return true, if both are equal.
+	 */
+	public boolean isEqualItemType( final ItemStack that, final ItemStack other )
+	{
+		if( that != null && other != null && that.getItem() == other.getItem() )
+		{
+			if( that.isItemStackDamageable() )
+			{
+				return true;
+			}
+			return that.getItemDamage() == other.getItemDamage();
+		}
+		return false;
+	}
+
+	/**
+	 * A wrapper around {@link ItemStack#isItemEqual(ItemStack)}.
+	 * 
+	 * The benefit is to compare two null item stacks, without any additional null checks.
+	 * 
+	 * Ignores NBT.
+	 * 
+	 * @return true, if both are equal.
+	 */
+	public boolean isEqualItem( @Nullable final ItemStack left, @Nullable final ItemStack right )
+	{
+		return left != null && right != null && left.isItemEqual( right );
+	}
+
+	/**
+	 * Compares two {@link ItemStack} and their NBT tag for equality.
+	 * 
+	 * Use this when a precise check is required and the same item is required.
+	 * Not just something with different NBT tags.
+	 * 
+	 * @return true, if both are identical.
+	 */
+	public boolean isSameItem( @Nullable final ItemStack is, @Nullable final ItemStack filter )
+	{
+		return isEqualItem( is, filter ) && hasSameNbtTag( is, filter );
+	}
+
+	/**
+	 * Similar to {@link ItemComparisonHelper#isEqualItem(ItemStack, ItemStack)},
+	 * but it can further check, if both match the same {@link FuzzyMode}
+	 * or are considered equal by the {@link OreDictionary}
+	 * 
+	 * @param mode how to compare the two {@link ItemStack}s
+	 * @return true, if both are matching the mode or considered equal by the {@link OreDictionary}
+	 */
+	public boolean isFuzzyEqualItem( final ItemStack a, final ItemStack b, final FuzzyMode mode )
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) The Cyclomatic Complexity of this method "isFuzzyEqualItem" is 24 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)